File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxDepth(TreeNode root) {
        // approach: recursive DFS
        
        return helper(root, 0);
    }
    
    private int helper(TreeNode node, int depth) {
        if (node == null) {
            return depth;
        }
        
        return Math.max(helper(node.left, depth + 1), helper(node.right, depth + 1));
    }
}

File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        # approach: top-down DFS, get max of left and right heights
        
        def helper(node, height):
            if not node:
                return height
            
            left = helper(node.left, height + 1)
            right = helper(node.right, height + 1)
            
            return max(left, right)
        
        return helper(root, 0)
            
            